skunkworks_crow/src/main/java/org/odk/share/activities/ScannerActivity.java

```diff
@@ -0,0 +1,122 @@
+import android.view.WindowManager;
+import android.widget.Button;
+
+import com.journeyapps.barcodescanner.CaptureManager;
+import com.journeyapps.barcodescanner.DecoratedBarcodeView;
+
+import org.odk.share.R;
+
+import androidx.annotation.NonNull;
+import androidx.appcompat.app.AppCompatActivity;
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import butterknife.OnClick;
+
+
+public class ScannerActivity extends AppCompatActivity implements
+        DecoratedBarcodeView.TorchListener {
+
+    private CaptureManager capture;
+    private boolean isFlashLightOn;
+    private boolean isFlashlightSupported;
+
+    @BindView(R.id.switch_flashlight)
+    Button toggleFlashlightButton;
+
+    @BindView(R.id.zxing_barcode_scanner)
+    DecoratedBarcodeView decoratedScannerView;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_scanner);
+        getWindow().addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
+
+        decoratedScannerView = findViewById(R.id.zxing_barcode_scanner);
+        decoratedScannerView.setTorchListener(this);
+        ButterKnife.bind(this);
+
+        if (!isFlashlightSupported()) {
+            toggleFlashlightButton.setVisibility(View.GONE);
+            isFlashlightSupported = false;
+        } else {
+            isFlashlightSupported = true;
+        }
+
+        capture = new CaptureManager(this, decoratedScannerView);
+        capture.initializeFromIntent(getIntent(), savedInstanceState);
+        capture.decode();
+    }
+
+    private boolean isFlashlightSupported() {
+        return getApplicationContext().getPackageManager()
+                .hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH);
+    }
+
+    public void toggleFlashlight() {
+        if (isFlashLightOn) {
+            decoratedScannerView.setTorchOff();
+            isFlashLightOn = false;
+        } else {
+            decoratedScannerView.setTorchOn();
+            isFlashLightOn = true;
+        }
+    }
+
+    @Override
+    public void onTorchOn() {
+        toggleFlashlightButton.setBackgroundResource(R.drawable.ic_flash_white_on);
+    }
+
+    @Override
+    public void onTorchOff() {
+        toggleFlashlightButton.setBackgroundResource(R.drawable.ic_flash_white_off);
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        capture.onResume();
+    }
+
+    @Override
+    protected void onPause() {
+        super.onPause();
+        capture.onPause();
+    }
+
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        capture.onDestroy();
+    }
+
+    @Override
+    protected void onSaveInstanceState(@NonNull Bundle outState) {
+        super.onSaveInstanceState(outState);
+        capture.onSaveInstanceState(outState);
+    }
+
+    @OnClick(R.id.switch_flashlight)
+    public void toggleButton() {
+        if (isFlashlightSupported) {
```

Comment:
This `if` condition will always be `true`. In `onCreate()`, you are setting the visibility to `GONE` if the feature isn't available. Hence, such a case where the button is clicked when the flashlight isn't available is not possible

Reply:
> This `if` condition will always be `true`. In `onCreate()`, you are setting the visibility to `GONE` if the feature isn't available. Hence, such a case where the button is clicked when the flashlight isn't available is not possible

Yeah, I didn't notice 🤪 , I updated it now